### PR TITLE
Proof of Concept for SASL plaintext auth

### PR DESCRIPTION
**Pull request by Crim**
for #105 

Proof of Concept showing SASL auth works.  This branch will evolve into a more complete solution.

### CODE REVIEW

**Comment by OlivierTarnus** [comment]
Really interested in this! When could we see a docker image. 

I'm happy to do some testing if this may help.

**Comment by Crim** [comment]
Hey @OlivierTarnus !  I'd love to get you to play around with it and provide some feedback.  Right now the easiest way would be to build and run the source code.  You can do that by checking out the source code in this branch and running the `./buildAndRun.sh` shell script, assuming you're running on *nix or OSX with java JDK 1.8+ installed.

If you absolutely need docker, I can likely cut a pre-release build that works on docker in another day or two?

Thanks!

File: kafka-webview-ui/src/main/java/org/sourcelab/kafka/webview/ui/controller/configuration/cluster/ClusterConfigController.java [hunk 1]
```diff
@@ -281,6 +301,46 @@ public String clusterUpdate(
             cluster.setTrustStorePassword(null);
         }
 
+        // If sasl is enabled
+        if (clusterForm.getSasl()) {
```
Comment (stanlemon) [inline kafka-webview-ui/src/main/java/org/sourcelab/kafka/webview/ui/controller/configuration/cluster/ClusterConfigController.java]: $.02 If you move the stuff up from below the else, you can probably figure out a way to short circuit here and remove a level of conditional nesting. Generally speaking this method stinks, it's big and hard to follow.

File: kafka-webview-ui/src/main/java/org/sourcelab/kafka/webview/ui/manager/kafka/KafkaConsumerFactory.java [hunk 2]
```diff
@@ -24,50 +24,37 @@
...
-     * @param keyStoreRootPath Parent path to where JKS key/trust stores are saved on disk.
-     * @param consumerIdPrefix Static prefix to pre-pend to all consumerIds.
+     * @param configUtil Utility class to DRY out common kafka client settings.
      */
-    public KafkaConsumerFactory(final String keyStoreRootPath, final String consumerIdPrefix) {
-        if (consumerIdPrefix == null) {
-            throw new IllegalArgumentException("ConsumerIdPrefix must be configured!");
+    public KafkaConsumerFactory(final KafkaClientConfigUtil configUtil) {
+        if (configUtil == null) {
+            throw new RuntimeException("Missing dependency KafkaClientConfigUtil!");
```
Comment (stanlemon) [inline kafka-webview-ui/src/main/java/org/sourcelab/kafka/webview/ui/manager/kafka/KafkaConsumerFactory.java]: Might consider a more specific exception that extends Runtime here.

File: kafka-webview-ui/src/main/java/org/sourcelab/kafka/webview/ui/manager/sasl/SaslProperties.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    /**
+     * Builder class for SaslProperties.
+     */
+    public static final class Builder {
```
Comment (stanlemon) [inline kafka-webview-ui/src/main/java/org/sourcelab/kafka/webview/ui/manager/sasl/SaslProperties.java]: Any reason not to split this out into it's own class available only to the package?

File: kafka-webview-ui/src/main/resources/schema/migration/h2/V1__init.sql [hunk 4]
```diff
@@ -0,0 +1,109 @@
+
+create table if not exists `user` (
```
Comment (stanlemon) [inline kafka-webview-ui/src/main/resources/schema/migration/h2/V1__init.sql]: Is this the flyway stuff? Does this get autogenerated or do you write this?

One note, for consistency sake you might want to match the casing on the other statements.

File: kafka-webview-ui/src/test/java/org/sourcelab/kafka/webview/ui/manager/kafka/KafkaAdminFactorySaslTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            // Call method to validate things work as expected
+            final DescribeClusterResult results = adminClient.describeCluster();
+            assertNotNull("Should have a non-null result", results);
+
+            // Request future result
+            final Collection<Node> nodes = results.nodes().get();
+            assertNotNull("Should have non-null node result", nodes);
+            assertFalse("Should have non-empty node", nodes.isEmpty());
+        }
+        //assertFalse("Finish writing this test", true);
```
Comment (stanlemon) [inline kafka-webview-ui/src/test/java/org/sourcelab/kafka/webview/ui/manager/kafka/KafkaAdminFactorySaslTest.java]: What's up with this?

**Comment by OlivierTarnus** [comment]
> Hey @OlivierTarnus ! I'd love to get you to play around with it and provide some feedback. Right now the easiest way would be to build and run the source code. You can do that by checking out the source code in this branch and running the `./buildAndRun.sh` shell script, assuming you're running on *nix or OSX with java JDK 1.8+ installed.
> 
> If you absolutely need docker, I can likely cut a pre-release build that works on docker in another day or two?
> 
> Thanks!

Hi,
We did a local build and I can confirm we successfully connected to our kafka cluster using SAL Plaintext.
I'll try to test Kerberos soon...

**Comment by Crim** [comment]
I've done manual testing and added automated tests over SSL, SASL_PLAIN, and SASL_SSL.  I haven't dug into Kerberos yet as I'll have to do some additional research to figure out how to set up a cluster configured with it. 

@OlivierTarnus If you're able to easily test/verify that, it'd be super helpful.  Otherwise I'll try to get into it soon.

**Comment by lgfa29** [comment]
Hi @Crim,

just wanted to chime in and say that I tested this branch agains an Event Streams Kafka cluster deployed on IBM Cloud which uses SASL_SSL and everything seems to be working.

Thanks for implementing this!
